File: backend/routes/logistics.py
from fastapi import APIRouter, HTTPException, Depends, Header
from pydantic import BaseModel
from typing import Optional, List, Dict
from datetime import datetime, timezone, timedelta
import jwt
import os
import random
import math

router = APIRouter(prefix="/api/logistics", tags=["logistics"])

db = None

def set_db(database):
    global db
    db = database

SECRET_KEY = os.environ.get("JWT_SECRET", "ocean-secret-key-2024")

async def verify_token(authorization: str = Header(None)):
    if not authorization:
        raise HTTPException(status_code=401, detail="Token مطلوب")
    try:
        token = authorization.replace("Bearer ", "")
        payload = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
        return payload
    except:
        raise HTTPException(status_code=401, detail="Token غير صالح")

# ==================== ROUTE OPTIMIZATION ====================

@router.get("/routes/optimize")
async def optimize_routes(driver_id: str = None, user = Depends(verify_token)):
    """تحسين المسارات"""
    return {
        "optimization_id": f"OPT-{random.randint(10000, 99999)}",
        "driver_id": driver_id or "all",
        "before": {
            "total_distance": 145,
            "total_time": "4:30 ساعات",
            "fuel_estimate": 25
        },
        "after": {
            "total_distance": 112,
            "total_time": "3:15 ساعات",
            "fuel_estimate": 19
        },
        "savings": {
            "distance": "33 كم (23%)",
            "time": "1:15 ساعة",
            "fuel": "6 لتر",
            "cost": "45 ر.س"
        },
        "optimized_route": [
            {"stop": 1, "order_id": "ORD-001", "address": "حي النخيل", "eta": "10:15"},
            {"stop": 2, "order_id": "ORD-005", "address": "حي النخيل", "eta": "10:25"},
            {"stop": 3, "order_id": "ORD-003", "address": "حي الملقا", "eta": "10:45"},
            {"stop": 4, "order_id": "ORD-002", "address": "حي الربيع", "eta": "11:05"},
            {"stop": 5, "order_id": "ORD-004", "address": "حي الربيع", "eta": "11:15"}
        ],
        "algorithm": "Genetic Algorithm + Google OR-Tools"
    }

@router.get("/routes/traffic")
async def get_traffic_conditions(user = Depends(verify_token)):
    """حالة المرور"""
    return {
        "zones": [
            {"name": "وسط الرياض", "status": "heavy", "delay": "15 دقيقة", "recommendation": "تجنب"},
            {"name": "شمال الرياض", "status": "moderate", "delay": "5 دقائق", "recommendation": "مقبول"},
            {"name": "جنوب الرياض", "status": "light", "delay": "0", "recommendation": "مثالي"},
            {"name": "شرق الرياض", "status": "moderate", "delay": "8 دقائق", "recommendation": "مقبول"}
        ],
        "incidents": [
            {"location": "طريق الملك فهد", "type": "حادث", "impact": "عالي", "expected_clear": "30 دقيقة"}
        ],
        "updated_at": datetime.now(timezone.utc).isoformat()
    }

# ==================== LIVE TRACKING ====================

@router.get("/tracking/{order_id}")
async def track_order(order_id: str, user = Depends(verify_token)):
    """تتبع الطلب الحي"""
    return {
        "order_id": order_id,
        "status": "جاري التوصيل",
        "driver": {
            "name": "أحمد محمد",
            "phone": "+966501234567",
            "vehicle": "تويوتا كامري - أبيض",
            "plate": "ABC 1234",
            "rating": 4.8
        },
        "location": {
            "lat": 24.7136 + random.uniform(-0.01, 0.01),
            "lng": 46.6753 + random.uniform(-0.01, 0.01),
            "updated_at": datetime.now(timezone.utc).isoformat()
        },
        "destination": {
            "lat": 24.7200,
            "lng": 46.6800,
            "address": "حي النخيل، الرياض"
        },
        "eta": "15 دقيقة",
        "distance_remaining": "4.2 كم",
        "timeline": [
            {"status": "تم الطلب", "time": "10:00", "completed": True},
            {"status": "جاري التجهيز", "time": "10:05", "completed": True},
            {"status": "تم الاستلام", "time": "10:20", "completed": True},
            {"status": "جاري التوصيل", "time": "10:25", "completed": True, "current": True},
            {"status": "تم التوصيل", "time": "10:45", "completed": False}
        ]
    }

@router.get("/tracking/fleet")
async def track_fleet(user = Depends(verify_token)):
    """تتبع الأسطول"""
    vehicles = []
    for i in range(20):
        vehicles.append({
            "id": f"VH-{i+1:03d}",
            "driver": f"سائق {i+1}",
            "status": random.choice(["delivering", "idle", "returning"]),
            "location": {
                "lat": 24.7136 + random.uniform(-0.1, 0.1),
                "lng": 46.6753 + random.uniform(-0.1, 0.1)
            },
            "current_order": f"ORD-{random.randint(10000, 99999)}" if random.random() > 0.3 else None,
            "speed": random.randint(0, 80),
            "battery": random.randint(20, 100)
        })
    return {"vehicles": vehicles, "total": len(vehicles)}

# ==================== SMART INVENTORY ====================

@router.get("/inventory/status")
async def get_inventory_status(user = Depends(verify_token)):
    """حالة المخزون"""
    return {
        "summary": {
            "total_skus": 5420,
            "healthy": 4850,
            "low_stock": 420,
            "out_of_stock": 85,
            "overstock": 65
        },
        "alerts": [
            {"product": "iPhone 15 Pro 256GB", "sku": "IP15P-256", "current": 12, "min": 50, "status": "critical"},
            {"product": "AirPods Pro 2", "sku": "APP2", "current": 25, "min": 40, "status": "low"},
            {"product": "Samsung S24 Ultra", "sku": "SS24U", "current": 350, "max": 200, "status": "overstock"}
        ],
        "auto_reorders_pending": 8,
        "warehouse_utilization": {
            "الرياض": 85,
            "جدة": 72,
            "الدمام": 68
        }
    }

@router.post("/inventory/reorder")
async def create_reorder(sku: str, quantity: int, supplier: str = None, user = Depends(verify_token)):
    """إنشاء طلب إعادة توريد"""
    return {
        "success": True,
        "reorder_id": f"RO-{random.randint(10000, 99999)}",
        "sku": sku,
        "quantity": quantity,
        "supplier": supplier or "مورد افتراضي",
        "estimated_arrival": (datetime.now(timezone.utc) + timedelta(days=random.randint(3, 7))).strftime("%Y-%m-%d"),
        "status": "pending"
    }

# ==================== DELIVERY SCHEDULING ====================

@router.get("/scheduling/slots")
async def get_delivery_slots(date: str = None, zone: str = None, user = Depends(verify_token)):
    """فترات التوصيل المتاحة"""
    target_date = date or datetime.now(timezone.utc).strftime("%Y-%m-%d")
    
    slots = [
        {"time": "09:00 - 11:00", "available": 15, "booked": 35, "price": 0},
        {"time": "11:00 - 13:00", "available": 8, "booked": 42, "price": 0},
        {"time": "13:00 - 15:00", "available": 22, "booked": 28, "price": 0},
        {"time": "15:00 - 17:00", "available": 5, "booked": 45, "price": 0},
        {"time": "17:00 - 19:00", "available": 3, "booked": 47, "price": 15},
        {"time": "19:00 - 21:00", "available": 0, "booked": 50, "price": 20},
        {"time": "21:00 - 23:00", "available": 12, "booked": 38, "price": 25}
    ]
    
    return {
        "date": target_date,
        "zone": zone or "جميع المناطق",
        "slots": slots,
        "express_available": True,
        "express_time": "30-45 دقيقة",
        "express_price": 35
    }

@router.post("/scheduling/book")
async def book_delivery_slot(order_id: str, date: str, slot: str, user = Depends(verify_token)):
    """حجز فترة توصيل"""
    return {
        "success": True,
        "booking_id": f"BK-{random.randint(10000, 99999)}",
        "order_id": order_id,
        "date": date,
        "slot": slot,
        "confirmed": True
    }

File: frontend/src/command-center/pages/Logistics.js
import React, { useState, useEffect } from 'react';

const Logistics = () => {
  const [activeTab, setActiveTab] = useState('routes');
  const [routeOptimization, setRouteOptimization] = useState(null);
  const [traffic, setTraffic] = useState(null);
  const [inventory, setInventory] = useState(null);
  const [deliverySlots, setDeliverySlots] = useState(null);
  const [fleet, setFleet] = useState([]);
  const [loading, setLoading] = useState(true);

  const API_URL = process.env.REACT_APP_BACKEND_URL;

  useEffect(() => {
    fetchData();
  }, [activeTab]);

  const fetchData = async () => {
    const token = localStorage.getItem('token');
    const headers = { 'Authorization': `Bearer ${token}` };

    try {
      if (activeTab === 'routes') {
        const [optRes, trafficRes] = await Promise.all([
          fetch(`${API_URL}/api/logistics/routes/optimize`, { headers }),
          fetch(`${API_URL}/api/logistics/routes/traffic`, { headers })
        ]);
        setRouteOptimization(await optRes.json());
        setTraffic(await trafficRes.json());
      } else if (activeTab === 'tracking') {
        const res = await fetch(`${API_URL}/api/logistics/tracking/fleet`, { headers });
        const data = await res.json();
        setFleet(data.vehicles || []);
      } else if (activeTab === 'inventory') {
        const res = await fetch(`${API_URL}/api/logistics/inventory/status`, { headers });
        setInventory(await res.json());
      } else if (activeTab === 'scheduling') {
        const res = await fetch(`${API_URL}/api/logistics/scheduling/slots`, { headers });
        setDeliverySlots(await res.json());
      }
    } catch (error) {
      console.error('Error:', error);
    } finally {
      setLoading(false);
    }
  };

  const tabs = [
    { id: 'routes', label: 'تحسين المسارات', icon: '🗝' },
    { id: 'tracking', label: 'تتبع الأسطول', icon: '🚚' },
    { id: 'inventory', label: 'المخزون الذكي', icon: '📦' },
    { id: 'scheduling', label: 'جدولة التوصيل', icon: '📅' }
  ];

  return (
    <div className="p-6 bg-gray-900 min-h-screen text-white" dir="rtl">
      <div className="mb-6">
        <h1 className="text-3xl font-bold mb-2">🚚 اللوجستيات الذكية</h1>
        <p className="text-gray-400">تحسين المسارات وإدارة الشحن</p>
      </div>

      {/* Tabs */}
      <div className="flex flex-wrap gap-2 mb-6">
        {tabs.map(tab => (
          <button
            key={tab.id}
            onClick={() => { setActiveTab(tab.id); setLoading(true); }}
            className={`px-4 py-2 rounded-lg transition-all ${activeTab === tab.id ? 'bg-orange-600 text-white' : 'bg-gray-800 text-gray-300 hover:bg-gray-700'}`}
          >
            {tab.icon} {tab.label}
          </button>
        ))}
      </div>

      {loading ? (
        <div className="flex justify-center items-center h-64">
          <div className="animate-spin rounded-full h-12 w-12 border-b-2 border-orange-500"></div>
        </div>
      ) : (
        <>
          {/* Routes Tab */}
          {activeTab === 'routes' && routeOptimization && (
            <div className="space-y-6">
              {/* Savings Summary */}
              <div className="bg-gradient-to-r from-green-900 to-teal-900 p-6 rounded-xl">
                <h3 className="text-xl font-bold mb-4">✨ نتائج التحسين</h3>
                <div className="grid grid-cols-2 md:grid-cols-4 gap-4">
                  <div>
                    <div className="text-green-200 text-sm">المسافة الموفرة</div>
                    <div className="text-2xl font-bold">{routeOptimization.savings?.distance}</div>
                  </div>
                  <div>
                    <div className="text-green-200 text-sm">الوقت الموفر</div>
                    <div className="text-2xl font-bold">{routeOptimization.savings?.time}</div>
                  </div>
                  <div>
                    <div className="text-green-200 text-sm">الوقود الموفر</div>
                    <div className="text-2xl font-bold">{routeOptimization.savings?.fuel}</div>
                  </div>
                  <div>
                    <div className="text-green-200 text-sm">التكلفة الموفرة</div>
                    <div className="text-2xl font-bold">{routeOptimization.savings?.cost}</div>
                  </div>
                </div>
              </div>

              {/* Route Details */}
              <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
                <div className="bg-gray-800 p-6 rounded-xl">
                  <h3 className="font-bold mb-4 text-red-400">❌ قبل التحسين</h3>
                  <div className="space-y-2">
                    <div className="flex justify-between"><span>المسافة:</span><span>{routeOptimization.before?.total_distance} كم</span></div>
                    <div className="flex justify-between"><span>الوقت:</span><span>{routeOptimization.before?.total_time}</span></div>
                    <div className="flex justify-between"><span>الوقود:</span><span>{routeOptimization.before?.fuel_estimate} لتر</span></div>
                  </div>
                </div>
                <div className="bg-gray-800 p-6 rounded-xl">
                  <h3 className="font-bold mb-4 text-green-400">✓ بعد التحسين</h3>
                  <div className="space-y-2">
                    <div className="flex justify-between"><span>المسافة:</span><span>{routeOptimization.after?.total_distance} كم</span></div>
                    <div className="flex justify-between"><span>الوقت:</span><span>{routeOptimization.after?.total_time}</span></div>
                    <div className="flex justify-between"><span>الوقود:</span><span>{routeOptimization.after?.fuel_estimate} لتر</span></div>
                  </div>
                </div>
              </div>

              {/* Optimized Route */}
              <div className="bg-gray-800 p-6 rounded-xl">
                <h3 className="font-bold mb-4">📍 المسار المحسّن</h3>
                <div className="flex flex-wrap gap-2">
                  {routeOptimization.optimized_route?.map((stop, idx) => (
                    <div key={idx} className="flex items-center gap-2">
                      <div className="bg-blue-600 w-8 h-8 rounded-full flex items-center justify-center font-bold">{stop.stop}</div>
                      <div className="bg-gray-700 p-2 rounded text-sm">
                        <div>{stop.address}</div>
                        <div className="text-gray-400 text-xs">{stop.order_id} - ETA: {stop.eta}</div>
                      </div>
                      {idx < routeOptimization.optimized_route.length - 1 && <span className="text-gray-500">→</span>}
                    </div>
                  ))}
                </div>
              </div>

              {/* Traffic */}
              {traffic && (
                <div className="bg-gray-800 p-6 rounded-xl">
                  <h3 className="font-bold mb-4">🚦 حالة المرور</h3>
                  <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4">
                    {traffic.zones?.map((zone, idx) => (
                      <div key={idx} className={`p-4 rounded-lg ${zone.status === 'light' ? 'bg-green-900/50' : zone.status === 'moderate' ? 'bg-yellow-900/50' : 'bg-red-900/50'}`}>
                        <div className="font-bold">{zone.name}</div>
                        <div className={`text-sm ${zone.status === 'light' ? 'text-green-400' : zone.status === 'moderate' ? 'text-yellow-400' : 'text-red-400'}`}>
                          {zone.status === 'light' ? 'سلس' : zone.status === 'moderate' ? 'متوسط' : 'مزدحم'}
                        </div>
                        {zone.delay !== '0' && <div className="text-xs text-gray-400">تأخير: {zone.delay}</div>}
                      </div>
                    ))}
                  </div>
                </div>
              )}
            </div>
          )}

          {/* Tracking Tab */}
          {activeTab === 'tracking' && (
            <div className="bg-gray-800 rounded-xl overflow-hidden">
              <div className="overflow-x-auto">
                <table className="w-full">
                  <thead className="bg-gray-700">
                    <tr>
                      <th className="p-3 text-right">المركبة</th>
                      <th className="p-3 text-right">السائق</th>
                      <th className="p-3 text-right">الحالة</th>
                      <th className="p-3 text-right">الطلب الحالي</th>
                      <th className="p-3 text-right">السرعة</th>
                      <th className="p-3 text-right">البطارية</th>
                    </tr>
                  </thead>
                  <tbody>
                    {fleet.slice(0, 15).map(v => (
                      <tr key={v.id} className="border-b border-gray-700 hover:bg-gray-750">
                        <td className="p-3 font-mono">{v.id}</td>
                        <td className="p-3">{v.driver}</td>
                        <td className="p-3">
                          <span className={`px-2 py-1 rounded text-xs ${v.status === 'delivering' ? 'bg-green-600' : v.status === 'idle' ? 'bg-blue-600' : 'bg-yellow-600'}`}>
                            {v.status === 'delivering' ? 'يوصل' : v.status === 'idle' ? 'متاح' : 'عائد'}
                          </span>
                        </td>
                        <td className="p-3">{v.current_order || '-'}</td>
                        <td className="p-3">{v.speed} كم/س</td>
                        <td className="p-3">
                          <div className="flex items-center gap-2">
                            <div className="w-16 bg-gray-600 rounded-full h-2">
                              <div className={`h-2 rounded-full ${v.battery > 50 ? 'bg-green-500' : v.battery > 20 ? 'bg-yellow-500' : 'bg-red-500'}`} style={{width: `${v.battery}%`}}></div>
                            </div>
                            <span className="text-xs">{v.battery}%</span>
                          </div>
                        </td>
                      </tr>
                    ))}
                  </tbody>
                </table>
              </div>
            </div>
          )}

          {/* Inventory Tab */}
          {activeTab === 'inventory' && inventory && (
            <div className="space-y-6">
              <div className="grid grid-cols-2 md:grid-cols-5 gap-4">
                <div className="bg-green-900 p-4 rounded-xl text-center">
                  <div className="text-2xl font-bold">{inventory.summary?.healthy}</div>
                  <div className="text-green-300 text-sm">سليم</div>
                </div>
                <div className="bg-yellow-900 p-4 rounded-xl text-center">
                  <div className="text-2xl font-bold">{inventory.summary?.low_stock}</div>
                  <div className="text-yellow-300 text-sm">منخفض</div>
                </div>
                <div className="bg-red-900 p-4 rounded-xl text-center">
                  <div className="text-2xl font-bold">{inventory.summary?.out_of_stock}</div>
                  <div className="text-red-300 text-sm">نافد</div>
                </div>
                <div className="bg-purple-900 p-4 rounded-xl text-center">
                  <div className="text-2xl font-bold">{inventory.summary?.overstock}</div>
                  <div className="text-purple-300 text-sm">فائض</div>
                </div>
                <div className="bg-blue-900 p-4 rounded-xl text-center">
                  <div className="text-2xl font-bold">{inventory.summary?.total_skus}</div>
                  <div className="text-blue-300 text-sm">إجمالي SKU</div>
                </div>
              </div>

              {/* Alerts */}
              <div className="bg-gray-800 p-6 rounded-xl">
                <h3 className="font-bold mb-4">🚨 تنبيهات المخزون</h3>
                <div className="space-y-3">
                  {inventory.alerts?.map((alert, idx) => (
                    <div key={idx} className={`p-4 rounded-lg border-r-4 ${alert.status === 'critical' ? 'bg-red-900/30 border-red-500' : alert.status === 'low' ? 'bg-yellow-900/30 border-yellow-500' : 'bg-purple-900/30 border-purple-500'}`}>
                      <div className="flex justify-between items-center">
                        <div>
                          <div className="font-bold">{alert.product}</div>
                          <div className="text-sm text-gray-400">SKU: {alert.sku}</div>
                        </div>
                        <div className="text-left">
                          <div className={`font-bold ${alert.status === 'overstock' ? 'text-purple-400' : 'text-red-400'}`}>
                            {alert.status === 'overstock' ? `فائض: ${alert.current}/${alert.max}` : `الحالي: ${alert.current} / الحد: ${alert.min}`}
                          </div>
                          <span className={`px-2 py-1 rounded text-xs ${alert.status === 'critical' ? 'bg-red-600' : alert.status === 'low' ? 'bg-yellow-600' : 'bg-purple-600'}`}>
                            {alert.status === 'critical' ? 'حرج' : alert.status === 'low' ? 'منخفض' : 'فائض'}
                          </span>
                        </div>
                      </div>
                    </div>
                  ))}
                </div>
              </div>

              {/* Warehouse Utilization */}
              <div className="bg-gray-800 p-6 rounded-xl">
                <h3 className="font-bold mb-4">🏭 استخدام المستودعات</h3>
                <div className="space-y-3">
                  {Object.entries(inventory.warehouse_utilization || {}).map(([wh, util]) => (
                    <div key={wh}>
                      <div className="flex justify-between text-sm mb-1">
                        <span>{wh}</span>
                        <span>{util}%</span>
                      </div>
                      <div className="w-full bg-gray-700 rounded-full h-3">
                        <div className={`h-3 rounded-full ${util > 90 ? 'bg-red-500' : util > 70 ? 'bg-yellow-500' : 'bg-green-500'}`} style={{width: `${util}%`}}></div>
                      </div>
                    </div>
                  ))}
                </div>
              </div>
            </div>
          )}

          {/* Scheduling Tab */}
          {activeTab === 'scheduling' && deliverySlots && (
            <div className="space-y-6">
              <div className="bg-gray-800 p-6 rounded-xl">
                <h3 className="font-bold mb-4">📅 فترات التوصيل - {deliverySlots.date}</h3>
                <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
                  {deliverySlots.slots?.map((slot, idx) => (
                    <div key={idx} className={`p-4 rounded-lg ${slot.available === 0 ? 'bg-red-900/30' : slot.available < 10 ? 'bg-yellow-900/30' : 'bg-green-900/30'}`}>
                      <div className="flex justify-between items-center">
                        <span className="font-bold">{slot.time}</span>
                        {slot.price > 0 && <span className="text-yellow-400 text-sm">+{slot.price} ر.س</span>}
                      </div>
                      <div className="mt-2 flex justify-between text-sm">
                        <span className={`${slot.available === 0 ? 'text-red-400' : 'text-green-400'}`}>
                          {slot.available === 0 ? 'ممتلئ' : `${slot.available} متاح`}
                        </span>
                        <span className="text-gray-400">{slot.booked} محجوز</span>
                      </div>
                      <div className="w-full bg-gray-700 rounded-full h-2 mt-2">
                        <div className="bg-blue-500 h-2 rounded-full" style={{width: `${(slot.booked / (slot.booked + slot.available)) * 100}%`}}></div>
                      </div>
                    </div>
                  ))}
                </div>
              </div>

              {deliverySlots.express_available && (
                <div className="bg-gradient-to-r from-orange-600 to-red-600 p-6 rounded-xl">
                  <div className="flex justify-between items-center">
                    <div>
                      <h3 className="text-xl font-bold">⚡ التوصيل السريع</h3>
                      <p className="text-orange-200">{deliverySlots.express_time}</p>
                    </div>
                    <div className="text-3xl font-bold">{deliverySlots.express_price} ر.س</div>
                  </div>
                </div>
              )}
            </div>
          )}
        </>
      )}
    </div>
  );
};

export default Logistics;